partiql-lang/src/main/resources/org/partiql/schemas/graph.isl

```diff
@@ -0,0 +1,71 @@
+$ion_schema_2_0
+
+/** This ISL schema is a specification of an Ion-based file format for defining graphs externally,
+    in order to be brought into PartiQL data model for querying with the PartQL implementation of GPML.
+    This is useful for general reasons of accessing external data, but is particularly necessary
+    for graphs, since PartiQL does not yet define concrete syntax for them.
+
+    The design of the schema closely follows the graph data model description in
+    PartiQL [RFC-0025](https://github.com/partiql/partiql-docs/blob/main/RFCs/0025-graph-data-model.md),
+    with the exception that the data model in this format is not recursive: the payloads at nodes and edges
+    are general Ion.  That is, it is not specified how a fragment of a payload can be itself recognized as a graph.
+
+    A file contains a graph if it contains a single Ion struct of ISL type Graph, as defined here.
+
+    Annotations specified in this schema (graph, node, edge, dir, undir) are for mnemonic purposes only --
+    graph authors are not required to use them and processors should not attribute semantic meaning to their
+    presence or absence.
+*/
+
+type::{
+    name: graph,  // ExternalGraphValidationTests breaks when this is named Graph ... no idea
+    annotations: closed::[graph],
+    type: struct,
+    fields: closed::{
+        nodes: { type: list, element: Node, occurs: required },
+        edges: { type: list, element: Edge, occurs: required }
+    }
+}
+
+type::{
+    name: Node,
+    annotations: closed::[node],
+    type: struct,
+    fields: closed::{
+        id: { type: NodeId, occurs: required },
+        labels: { type: list, element: string, occurs: optional },
+        payload: { type: NodePayload, occurs: optional }
+    }
+}
+
+type::{
+    name: Edge,
+    annotations: closed::[edge],
+    type: struct,
+    fields: closed::{
+        id: { type: EdgeId, occurs: required },
+        labels: { type: list, element: string},
+        ends: { one_of: [DirectedPair, UndirectedPair], occurs: required },
+        payload: { type: EdgePayload, occurs: optional }
+    }
+}
```

Comment:
This pattern of a product wrapping a single union type field has come up pretty often in protobuf, isl, pig, sprout, and I'm sure many more models. You have variants with some shared fields, so you create an outer product time called "common" and the variants are in something called "body". In this case we have `Edge` and `Ends variants DirectedPair, UndirectedPair`.

In a perfect world, we would define an Edge as a union type with common properties (id, labels, ends) amongst variants but we don't have the modeling power so we have to do this product-wrapped-union thing.

Purely aesthetics, but I've found the name for the "body" union type to not carry much meaning and can be dropped without suffering. In Ion, we can use `_` as an "empty" symbol, so your `ends` can be removed. Also moving required fields first. Then we get

```ion
type::{
    name: Edge,
    annotations: closed::[edge],
    type: struct,
    fields: closed::{
        id: { type: EdgeId, occurs: required },
        _: { one_of: [DirectedPair, UndirectedPair], occurs: required }
        labels: { type: list, element: string},
        payload: { type: EdgePayload, occurs: optional }
    }
}
```

And we can have edges like
```
edge::{ id: a, _: ( x -> y ) }
```

Reply:
I feel like passing on the points here at the moment...

Overall, the schema is simple enough that factoring out the commonality between nodes and edges could only increase understanding burden for someone who needs to write a graph in a hurry, I feel.  

As for `ends` vs `_` -- I am inclined to be explicit, a.k.a "self-describing".